FastSimulation/SimplifiedGeometryPropagator/src/HelixTrajectory.cc

```diff
@@ -0,0 +1,212 @@
+    //, centerX_(position_.X() - radius_*std::cos(phi_))
+    //, centerY_(position_.Y() - radius_*std::sin(phi_))
+    , centerR_(std::sqrt(centerX_*centerX_ + centerY_*centerY_))
+    , minR_(std::abs(centerR_ - radius_))
+    , maxR_(centerR_ + radius_)
+    // omega = q * e * B / (gamma * m) = q * e *B / (E / c^2) = q * e * B * c^2 / E
+    // omega: negative for negative q -> seems to be what we want.
+    // energy in units of GeV: omega = q * B * c^2 / (E * 10^9)
+    // in cmssw units: omega[1/ns] = q * B * c^2 * 10^-4 / E
+    , phiSpeed_(-particle.charge() * magneticFieldZ * fastsim::Constants::speedOfLight * fastsim::Constants::speedOfLight * 1e-4 / momentum_.E())
+{;}
+
+bool fastsim::HelixTrajectory::crosses(const BarrelSimplifiedGeometry & layer) const
+{
+    return (minR_ < layer.getRadius() && maxR_ > layer.getRadius());
+}
+
+double fastsim::HelixTrajectory::nextCrossingTimeC(const BarrelSimplifiedGeometry & layer, bool onLayer) const
+{
+    if(!crosses(layer)) return -1;
+
+    // solve the following equation for sin(phi)
+    // (x^2 + y^2 = R_L^2)     (1)      the layer 
+    // x = x_c + R_H*cos(phi)  (2)      the helix in the xy plane
+    // y = y_c + R_H*sin(phi)  (3)      the helix in the xy plane
+    // with
+    // R_L: the radius of the layer
+    // x_c,y_c the center of the helix in xy plane
+    // R_H, the radius of the helix
+    // phi, the phase of the helix
+    //
+    // substitute (2) and (3) in (1)
+    // =>
+    //   x_c^2 + 2*x_c*R_H*cos(phi) + R_H^2*cos^2(phi)
+    // + y_c^2 + 2*y_c*R_H*sin(phi) + R_H^2*sin^2(phi)
+    // = R_L^2
+    // =>
+    // (x_c^2 + y_c^2 + R_H^2 - R_L^2) + (2*y_c*R_H)*sin(phi) = -(2*x_c*R_H)*cos(phi)
+    //
+    // rewrite
+    //               E                 +       F    *sin(phi) =      G     *cos(phi)
+    // =>
+    // E^2 + 2*E*F*sin(phi) + F^2*sin^2(phi) = G^2*(1-sin^2(phi))
+    // rearrange
+    // sin^2(phi)*(F^2 + G^2) + sin(phi)*(2*E*F) + (E^2 - G^2) = 0
+    //
+    // rewrite
+    // sin^2(phi)*     a      + sin(phi)*   b    +      c      = 0
+    // => sin(phi) = (-b +/- sqrt(b^2 - 4*ac)) / (2*a)
+    // with
+    // a = F^2 + G^2
+    // b = 2*E*F
+    // c = E^2 - G^2
+
+    double E = centerX_*centerX_ + centerY_*centerY_ + radius_*radius_ - layer.getRadius()*layer.getRadius();
+    double F = 2*centerY_*radius_;
+    double G = 2*centerX_*radius_;
+
+    double a = F*F + G*G;
+    double b = 2*E*F;
+    double c = E*E - G*G;
+
+    double delta = b*b - 4*a*c;
+
+    // case of no solution
+    if(delta < 0)
+    {   
+        // Should not be reached: Full Propagation does always have a solution "if(crosses(layer)) == -1"
+        // Even if particle is outside all layers -> can turn around in magnetic field
+        throw cms::Exception("FastSimulation") << "HelixTrajectory: should not be reached (no solution).";
```

Comment:
@skurz, since you are checking the effect of this already merged PR: perhaps the following failure 
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22152/26029/addOnTests/fastsim2/cmsDriver.py_TTbar_13TeV_TuneCUETP8M1_cfi_--conditions_auto:run2_mc_--fast__-n_100_--eventcontent_AODSIM,DQM_--relval_100000,1000_-s_GEN,SIM,RECOBEFMI.log
observed in the addOn tests of the apparently unrelated pull request #22152 is also worth giving a look

Reply:
Yes, this is definitely related to my PR but still it is really strange. I'll try to find out what is going on.